apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/bulk/ZstdGroupedWeight.scala

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package org.lfdecentralizedtrust.splice.scan.store.bulk
+
+import com.github.luben.zstd.ZstdDirectBufferCompressingStreamNoFinalizer
+import org.apache.pekko.stream.{Attributes, FlowShape, Inlet, Outlet}
+import org.apache.pekko.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import org.apache.pekko.util.ByteString
+
+import java.nio.ByteBuffer
+import java.util.concurrent.atomic.AtomicReference
+
+/** A Pekko GraphStage that zstd-compresses a stream of bytestrings, and splits the output into zstd objects of size (minWeight + delta).
+  * Somewhat similar to Pekko's built-in GroupedWeight, but outputs valid zstd compressed objects.
+  */
+case class ZstdGroupedWeight(minSize: Long) extends GraphStage[FlowShape[ByteString, ByteString]] {
+  require(minSize > 0, "minSize must be greater than 0")
+
+  val zstdTmpBufferSize = 10 * 1024 * 1024; // TODO(#3429): make configurable?
+
+  val in = Inlet[ByteString]("ZstdGroupedWeight.in")
+  val out = Outlet[ByteString]("ZstdGroupedWeight.out")
+  override val shape: FlowShape[ByteString, ByteString] = FlowShape(in, out)
+
+  override def initialAttributes: Attributes = Attributes.name("ZstdGroupedWeight")
+
+  private case class State(
+      bytes: ByteString,
+      left: Long,
+  ) {
+    def append(bs: ByteString): State =
+      State(
+        bytes ++ bs,
+        left - bs.size,
+      )
+
+  }
+  private object State {
+    def empty(): State = State(ByteString.empty, minSize)
+  }
+
+  class ZSTD(
+      val tmpBuffer: ByteBuffer,
+      val compressionLevel: Int = 3,
+  ) extends AutoCloseable {
+
+    val compressingStream =
+      new ZstdDirectBufferCompressingStreamNoFinalizer(tmpBuffer, compressionLevel)
+
+    def compress(input: ByteString): ByteString = {
+      val inputBB = ByteBuffer.allocateDirect(input.size)
```

Comment:
can you not reuse this buffer instead of creating a new one for every compression? allocateDirect is slow, GC does not see native mem so it can take a while before it gets cleaned up, in the meantime native memory starts piling up. maybe better, since you always use this from graphstagelogic, is to have a reusable direct buffer there, pass it into this method, and reuse it. (otherwise you risk running out of memory by piling up native memory while gc is not necessary, if you start streaming a lot of files)

Reply:
https://openjdk.org/jeps/442 Project Panama in JDK 22 should make things a lot better, but likely that still would not work with this library, they'll need to update it, or we of course could consider contributing...